File: backend/models/project_knowledge.py
"""Data contracts for project knowledge, Reviewer proposals, and file plans."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal, Optional
from uuid import uuid4

from pydantic import BaseModel, Field, model_validator


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


KnowledgeType = Literal[
    "finding",
    "conclusion",
    "decision",
    "hypothesis",
    "question",
    "task",
    "project_change",
    "artifact",
]
Confidence = Literal["low", "medium", "high"]
Importance = Literal["normal", "important", "critical"]
ProposalStatus = Literal["pending", "accepted", "rejected", "failed", "undone"]
FileOperationType = Literal["mkdir", "move", "rename"]


class SourceReference(BaseModel):
    session_id: Optional[str] = None
    message_ids: list[str] = Field(default_factory=list)
    files: list[str] = Field(default_factory=list)
    run_ids: list[str] = Field(default_factory=list)
    citations: list[str] = Field(default_factory=list)


class FileOperation(BaseModel):
    type: FileOperationType
    source: Optional[str] = None
    target: str = Field(min_length=1, max_length=1000)
    reason: str = Field(default="", max_length=2000)

    @model_validator(mode="after")
    def validate_source(self):
        if self.type in {"move", "rename"} and not self.source:
            raise ValueError(f"{self.type} requires source")
        if self.type == "mkdir" and self.source:
            raise ValueError("mkdir does not accept source")
        return self


class ReviewerProposalInput(BaseModel):
    proposal_type: Literal["knowledge", "file_operation"]
    knowledge_type: Optional[KnowledgeType] = None
    title: str = Field(min_length=1, max_length=160)
    summary: str = Field(min_length=1, max_length=4000)
    reason: str = Field(min_length=1, max_length=2000)
    confidence: Confidence = "medium"
    importance: Importance = "normal"
    source_message_ids: list[str] = Field(default_factory=list, max_length=100)
    related_files: list[str] = Field(default_factory=list, max_length=100)
    conflicts_with: list[str] = Field(default_factory=list, max_length=100)
    supersedes: list[str] = Field(default_factory=list, max_length=100)
    operations: list[FileOperation] = Field(default_factory=list, max_length=50)

    @model_validator(mode="after")
    def validate_shape(self):
        if self.proposal_type == "knowledge":
            if not self.knowledge_type:
                raise ValueError("knowledge proposal requires knowledge_type")
            if self.operations:
                raise ValueError("knowledge proposal cannot include file operations")
        else:
            if self.knowledge_type:
                raise ValueError("file operation proposal cannot include knowledge_type")
            if not self.operations:
                raise ValueError("file operation proposal requires operations")
        return self


class ReviewerResult(BaseModel):
    proposals: list[ReviewerProposalInput] = Field(default_factory=list, max_length=20)


class Proposal(ReviewerProposalInput):
    id: str = Field(default_factory=lambda: f"proposal-{uuid4().hex[:12]}")
    status: ProposalStatus = "pending"
    source: SourceReference = Field(default_factory=SourceReference)
    fingerprint: str = ""
    reviewer_run_id: Optional[str] = None
    created_at: str = Field(default_factory=utc_now_iso)
    updated_at: str = Field(default_factory=utc_now_iso)
    decision_reason: Optional[str] = None
    applied_history_id: Optional[str] = None


class KnowledgeItem(BaseModel):
    id: str = Field(default_factory=lambda: f"knowledge-{uuid4().hex[:12]}")
    type: KnowledgeType
    title: str = Field(min_length=1, max_length=160)
    summary: str = Field(min_length=1, max_length=4000)
    confidence: Confidence = "medium"
    importance: Importance = "normal"
    status: Literal["active", "superseded", "archived"] = "active"
    source: SourceReference = Field(default_factory=SourceReference)
    related_files: list[str] = Field(default_factory=list)
    conflicts_with: list[str] = Field(default_factory=list)
    supersedes: list[str] = Field(default_factory=list)
    proposal_id: Optional[str] = None
    created_at: str = Field(default_factory=utc_now_iso)
    updated_at: str = Field(default_factory=utc_now_iso)


class ProjectPolicy(BaseModel):
    auto_review: bool = True
    reminder_threshold: int = Field(default=5, ge=1, le=100)
    max_directory_depth: int = Field(default=3, ge=1, le=10)
    minimum_files_for_new_category: int = Field(default=3, ge=1, le=100)
    locked_paths: list[str] = Field(default_factory=list)
    naming_pattern: str = "{date}_{topic}_{kind}_{version}"
    accepted_counts: dict[str, int] = Field(default_factory=dict)
    rejected_counts: dict[str, int] = Field(default_factory=dict)
    external_services_allowed: bool = True
    allowed_egress_domains: list[str] = Field(default_factory=list)
    blocked_data_classes: list[str] = Field(default_factory=list)
    updated_at: str = Field(default_factory=utc_now_iso)


class ReviewRequest(BaseModel):
    cwd: str
    session_id: Optional[str] = None
    include_files: bool = True
    force_full_session: bool = False


class ReviewResponse(BaseModel):
    ok: bool = True
    run_id: str
    created: int
    skipped: int = 0
    proposal_ids: list[str] = Field(default_factory=list)
    message: str = ""


class ProposalDecisionRequest(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=160)
    summary: Optional[str] = Field(default=None, min_length=1, max_length=4000)
    reason: Optional[str] = Field(default=None, max_length=2000)


class ProposalUpdateRequest(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=160)
    summary: Optional[str] = Field(default=None, min_length=1, max_length=4000)
    reason: Optional[str] = Field(default=None, min_length=1, max_length=2000)
    confidence: Optional[Confidence] = None
    importance: Optional[Importance] = None
    related_files: Optional[list[str]] = None
    operations: Optional[list[FileOperation]] = None


class BatchDecisionRequest(BaseModel):
    proposal_ids: list[str] = Field(min_length=1, max_length=100)
    action: Literal["accept", "reject"]
    reason: Optional[str] = Field(default=None, max_length=2000)


class ProjectPolicyUpdate(BaseModel):
    auto_review: Optional[bool] = None
    reminder_threshold: Optional[int] = Field(default=None, ge=1, le=100)
    max_directory_depth: Optional[int] = Field(default=None, ge=1, le=10)
    minimum_files_for_new_category: Optional[int] = Field(default=None, ge=1, le=100)
    locked_paths: Optional[list[str]] = None
    naming_pattern: Optional[str] = Field(default=None, max_length=200)
    external_services_allowed: Optional[bool] = None
    allowed_egress_domains: Optional[list[str]] = None
    blocked_data_classes: Optional[list[str]] = None

File: frontend/src/lib/project-knowledge.ts
import { useQuery } from "@tanstack/react-query";
import { apiRequest } from "./api";
import { queryClient } from "./query-client";

export type KnowledgeType =
  | "finding"
  | "conclusion"
  | "decision"
  | "hypothesis"
  | "question"
  | "task"
  | "project_change"
  | "artifact";

export type ProposalStatus = "pending" | "accepted" | "rejected" | "failed" | "undone";

export interface SourceReference {
  session_id?: string | null;
  message_ids: string[];
  files: string[];
  run_ids: string[];
  citations: string[];
}

export interface FileOperation {
  type: "mkdir" | "move" | "rename";
  source?: string | null;
  target: string;
  reason?: string;
}

export interface Proposal {
  id: string;
  proposal_type: "knowledge" | "file_operation";
  knowledge_type?: KnowledgeType | null;
  title: string;
  summary: string;
  reason: string;
  confidence: "low" | "medium" | "high";
  importance: "normal" | "important" | "critical";
  source_message_ids: string[];
  related_files: string[];
  conflicts_with: string[];
  supersedes: string[];
  operations: FileOperation[];
  experience_ids?: string[];
  loop_ids?: string[];
  candidate_ids?: string[];
  evaluator_refs?: Array<Record<string, unknown>>;
  artifact_refs?: Array<Record<string, unknown>>;
  status: ProposalStatus;
  source: SourceReference;
  created_at: string;
  updated_at: string;
  decision_reason?: string | null;
  applied_history_id?: string | null;
}

export interface KnowledgeItem {
  id: string;
  type: KnowledgeType;
  title: string;
  summary: string;
  confidence: "low" | "medium" | "high";
  importance: "normal" | "important" | "critical";
  status: "active" | "superseded" | "archived";
  source: SourceReference;
  related_files: string[];
  conflicts_with: string[];
  supersedes: string[];
  experience_ids?: string[];
  loop_ids?: string[];
  candidate_ids?: string[];
  evaluator_refs?: Array<Record<string, unknown>>;
  artifact_refs?: Array<Record<string, unknown>>;
  created_at: string;
  updated_at: string;
}

export interface ProjectSummary {
  workspace: string;
  project_file: string;
  pending_count: number;
  knowledge_count: number;
  auto_review: boolean;
}

export interface ProjectPolicy {
  auto_review: boolean;
  reminder_threshold: number;
  max_directory_depth: number;
  minimum_files_for_new_category: number;
  locked_paths: string[];
  naming_pattern: string;
  accepted_counts: Record<string, number>;
  rejected_counts: Record<string, number>;
  updated_at: string;
}

export interface IndexedFile {
  id: string;
  path: string;
  name: string;
  directory: string;
  extension: string;
  kind: string;
  size: number;
  modified: number;
  fingerprint: string;
  tags: string[];
}

export interface LogicalFileViews {
  updated_at: string;
  files: IndexedFile[];
  by_type: Record<string, IndexedFile[]>;
  by_topic: Record<string, IndexedFile[]>;
  by_month: Record<string, IndexedFile[]>;
}

/** Every project-knowledge read shares this prefix so one mutation invalidates them all. */
export const projectKnowledgeKey = (...selector: Array<string | null>) => ["project-knowledge", ...selector];

// This resource kept a 5s TTL where the rest of the app used 3s; preserved here.
const KNOWLEDGE_STALE_MS = 5_000;

function read<T>(queryKey: Array<string | null>, path: string) {
  return { queryKey, queryFn: () => apiRequest<T>(path), staleTime: KNOWLEDGE_STALE_MS };
}

/** Imperative read for non-component callers; shares the cache with the hooks below. */
function get<T>(queryKey: Array<string | null>, path: string): Promise<T> {
  return queryClient.fetchQuery(read<T>(queryKey, path));
}

/** Writes go straight to the transport, then drop the whole resource from cache. */
async function write<T>(path: string, init: RequestInit): Promise<T> {
  const data = await apiRequest<T>(path, init);
  void queryClient.invalidateQueries({ queryKey: projectKnowledgeKey() });
  return data;
}

function query(cwd: string, extra?: Record<string, string>) {
  return new URLSearchParams({ cwd, ...(extra ?? {}) }).toString();
}

const proposalCountQuery = (cwd: string) => read<{ pending_count: number }>(projectKnowledgeKey("proposals-count", cwd), `/api/project-knowledge/proposals/count?${query(cwd)}`);

const fileViewsQuery = (cwd: string) => read<LogicalFileViews>(projectKnowledgeKey("file-views", cwd), `/api/project-knowledge/files/views?${query(cwd)}`);

/** Logical file views: mounted only by the files tab, refreshed by the write invalidation. */
export function useLogicalFileViews(cwd: string) {
  return useQuery(fileViewsQuery(cwd));
}

/** Pending-proposal badge: polled, not pushed — the server has no signal for it. */
export function usePendingProposalCount(cwd: string, refetchIntervalMs: number) {
  return useQuery({ ...proposalCountQuery(cwd), refetchInterval: refetchIntervalMs });
}

export const projectKnowledgeApi = {
  summary(cwd: string) {
    return get<ProjectSummary>(projectKnowledgeKey("summary", cwd), `/api/project-knowledge/summary?${query(cwd)}`);
  },
  project(cwd: string) {
    return get<ProjectSummary & { content: string }>(projectKnowledgeKey("project", cwd), `/api/project-knowledge/project?${query(cwd)}`);
  },
  projectVersions(cwd: string) {
    return get<{ versions: Array<{ id: string; created_at: string; reason: string; knowledge_count: number }> }>(projectKnowledgeKey("project-versions", cwd), `/api/project-knowledge/project/versions?${query(cwd)}`);
  },
  restoreProjectVersion(cwd: string, versionId: string) {
    return write<Record<string, unknown>>(`/api/project-knowledge/project/versions/${versionId}/restore?${query(cwd)}`, {
      method: "POST",
    });
  },
  proposals(cwd: string, status?: ProposalStatus) {
    const extra = status ? { status } : undefined;
    return get<{ proposals: Proposal[]; pending_count: number }>(projectKnowledgeKey("proposals", cwd, status ?? null), `/api/project-knowledge/proposals?${query(cwd, extra)}`);
  },
  proposalCount(cwd: string) {
    return queryClient.fetchQuery(proposalCountQuery(cwd));
  },
  items(cwd: string) {
    return get<{ items: KnowledgeItem[] }>(projectKnowledgeKey("items", cwd), `/api/project-knowledge/items?${query(cwd)}`);
  },
  review(cwd: string, sessionId?: string | null, forceFullSession = false) {
    return write<{ run_id: string; created: number; skipped: number; proposal_ids: string[]; message: string }>(
      "/api/project-knowledge/review",
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          cwd,
          session_id: sessionId || null,
          include_files: true,
          force_full_session: forceFullSession,
        }),
      },
    );
  },
  updateProposal(cwd: string, proposalId: string, changes: Partial<Proposal>) {
    return write<{ proposal: Proposal }>(`/api/project-knowledge/proposals/${proposalId}?${query(cwd)}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(changes),
    });
  },
  previewProposal(cwd: string, proposalId: string) {
    return get<Record<string, unknown>>(projectKnowledgeKey("proposal-preview", cwd, proposalId), `/api/project-knowledge/proposals/${proposalId}/preview?${query(cwd)}`);
  },
  accept(cwd: string, proposalId: string, edits?: { title?: string; summary?: string }) {
    return write<Record<string, unknown>>(`/api/project-knowledge/proposals/${proposalId}/accept?${query(cwd)}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(edits ?? {}),
    });
  },
  reject(cwd: string, proposalId: string, reason?: string) {
    return write<Record<string, unknown>>(`/api/project-knowledge/proposals/${proposalId}/reject?${query(cwd)}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ reason }),
    });
  },
  batch(cwd: string, proposalIds: string[], action: "accept" | "reject") {
    return write<{ ok: boolean; failures: Array<{ proposal_id: string; detail: string }> }>(
      `/api/project-knowledge/proposals/batch?${query(cwd)}`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ proposal_ids: proposalIds, action }),
      },
    );
  },
  policy(cwd: string) {
    return get<ProjectPolicy>(projectKnowledgeKey("policy", cwd), `/api/project-knowledge/policy?${query(cwd)}`);
  },
  updatePolicy(cwd: string, changes: Partial<ProjectPolicy>) {
    return write<ProjectPolicy>(`/api/project-knowledge/policy?${query(cwd)}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(changes),
    });
  },
  files(cwd: string) {
    return queryClient.fetchQuery(fileViewsQuery(cwd));
  },
  history(cwd: string) {
    return get<{ history: Array<Record<string, unknown>> }>(projectKnowledgeKey("history", cwd), `/api/project-knowledge/history?${query(cwd)}`);
  },
  undo(cwd: string, historyId: string) {
    return write<Record<string, unknown>>(`/api/project-knowledge/file-operations/${historyId}/undo?${query(cwd)}`, {
      method: "POST",
    });
  },
};

export const KNOWLEDGE_LABELS: Record<KnowledgeType, string> = {
  finding: "Finding",
  conclusion: "Conclusion",
  decision: "Decision",
  hypothesis: "Hypothesis",
  question: "Open question",
  task: "Next step",
  project_change: "Project change",
  artifact: "Artifact",
};

export function groupKnowledgeItems(items: KnowledgeItem[]) {
  return items.reduce<Record<string, KnowledgeItem[]>>((groups, item) => {
    (groups[item.type] ??= []).push(item);
    return groups;
  }, {});
}

export function formatFileSize(bytes: number) {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  if (bytes < 1024 * 1024 * 1024) return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
  return `${(bytes / (1024 * 1024 * 1024)).toFixed(1)} GB`;
}
